### ISSUE TITLE
 "net/http: request canceled while waiting for connection"

### ISSUE THREAD

**Issue by aranhoide** [getlantern/lantern#2394]
This is the most frequent error log at loggly at the time of this writing.

**Comment by fffw** [comment]
From loggly, those errors are reported from two packages --- `flashlight.statserver` and `flashlight.geolookup`, but triggered from single line of code: [geolookup.LookupIPWithClient](https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/geolookup/geolookup.go#L122)

**Comment by fffw** [comment]
`geolookup` uses a `http.Client` with 20 seconds timeout to access geoserver. It will `CancelRequest` after 20 seconds. [`http.Transport`](http://golang.org/src/net/http/transport.go#L539) will report this error as a consequence.

Other clients, including Loggly client and cloud config fetcher, use `HTTPClient` from `flashlight/util` without a timeout, so this error will not occur to them.

We may need to remove this timeout and compare the error count of geolookup versus fetching cloud config.

**Comment by fffw** [comment]
The intervals of fetching cloud config (1 min) and doing geolookup (30 secs) are too small. They don't change often, continuously pulling is a waste of bandwidth. I suggest an interval of 30 min - 1 hour or even longer.

**Comment by fffw** [comment]
Oh sorry @xiam I just noticed that this issue was assigned to you! We may discuss a bit about it.

**Comment by xiam** [comment]
Don't worry @fffw, I've been not able to reproduce it. We may as well keep on with the discussion.

**Comment by myleshorton** [comment]
Hey @fffw I actually think we should probably include *some* timeout and should add a timeout to the other services. Otherwise we'll just be left with idle connections that can ultimately contribute to file descriptor limit issues. I would just make the timeouts really high, like 300 seconds.

**Comment by fffw** [comment]
I'm not sure enough but I'll check if dialer and enproxy have timeout themselves. Anyway, 20 seconds is long enough to establish connection so we should dig deeper on why the error is reported so frequently.

**Comment by xiam** [comment]
It seems like the sleep time before retrying to geolocate a peer in case of error was too small, also we were missing some details in the log.

> I would just make the timeouts really high, like 300 seconds.

@myleshorton I think 300s is too much, the error we were seeing was just the last error and we can't be sure if every request before that last one timed out too. Instead of trying with 300s I proposed a more conservative increment of 60s, put more time between attempts and added more details (attempt number and error message) to the log.

Please see this PR https://github.com/getlantern/lantern/pull/2474

**Comment by fffw** [comment]
I suspect the frequent timeout is because of the hqfd (fronted) dialer geolookup using, see https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/flashlight.go#L247

I tried marking all chained servers as untrusted and `http_proxy=localhost:8787 time curl http://geo.getiantem.org/lookup/1.1.1.1`, it hit nl.fallbacks.getiantem.org and took 10+ seconds to complete, whereas it only took less than 1s via chained servers and jp.fallbacks.getiantem.org. I wanted to catch a 30s+ RTT but didn't succeeded. Anyway, limiting geolookup to fronted servers actually *degrades* QoS in this case.

Loggly and cloud config fetcher simply uses `util.HTTPClient`. That may explains why we rarely saw timeouts for them.

**Comment by myleshorton** [comment]
Interesting @fffw thanks. So just to make sure we're all on the same page, util.HTTPClient will work because it hits Lantern as a local proxy, which then goes through all the normal proxying logic and in practice ends up using a chained proxy typically.

Actually, come to think of it, how will that interact with detour? I.e. it will likely go directly if detour detects loggly and amazon as unblocked correct?

That geolookup ends up being very important because it determines other things like the country-specific cloud.yaml to lookup. It taking a long time currently also delays the startup of the local proxy, which will actually cause loggly and the auto-updates to fail because they time out waiting for the local server after 5 seconds!

**Comment by myleshorton** [comment]
The original intent there was to use fronted servers instead of peer servers (when Lantern 2.0 was domain fronting-only). This issue relates to #2321, but I'm basically wondering if it makes sense to just use ```util.HTTPClient``` instead at https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/flashlight.go#L245

Is that what you're thinking? Right now we have no peers, but when we do I think we only want to use ones that are super reliable.

**Comment by fffw** [comment]
Currently, use `util.HTTPClient` may reduce the frequency of this error I think. Current problem is that the QoS in yaml is not real. We can measure the actual QoS of a server once we implement sever selection in client. Before then, can we just assign same QoS to all servers?

**Comment by myleshorton** [comment]
Yeah that's a good idea @fffw. That'll allow us to reduce the QoS of peers as well. You want to ticket it for beta 7?

**Comment by fffw** [comment]
They are already the same! See https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/genconfig/cloud.yaml.tmpl

Good news is, I never saw this "request canceled" error again since I removed old lantern.yaml yestoday. Unfortunately I didn't keep the old one to show the difference.

**Comment by myleshorton** [comment]
Hey @fffw I just took a look at https://s3.amazonaws.com/lantern_config/cloud.2.0.0-nl.yaml.gz, and we actually are assigning the same QoS to chained servers and fronted servers. I'm wondering if there's something else going on. We also could assign chained servers a *higher QoS* than fronted servers. What do you think?

**Comment by myleshorton** [comment]
Oh sorry -- I commented without refreshing the browser window, so I missed your comment =). 

I've actually seen that error in @sheena3's logs, which led me back down this same road. I do think we should:

1. Just use util.HTTPClient for all those services
1. Assign chained proxies a higher QoS.

What do you think?

**Comment by xiam** [comment]
I'm still seeing recent logs on this issue.

@myleshorton I think that's a good idea, I'm going to try `util.HTTPClient()` and adjust some waiting times in order to make them longer.

**Comment by xiam** [comment]
I just tried and it gave me this result:

![screen shot 2015-06-08 at 4 08 49 pm](https://cloud.githubusercontent.com/assets/385670/8045618/2aa87a28-0dfa-11e5-9fb4-efbdbac27d79.png)

The actual reply I'm seeing from the geolookup service is this:

```
{"City":{"GeoNameID":0,"Names":null},"Continent":{"Code":"","GeoNameID":0,"Names":null},"Country":{"GeoNameID":0,"IsoCode":"","Names":null},"Location":{"Latitude":0,"Longitude":0,"MetroCode":0,"TimeZone":""},"Postal":{"Code":""},"RegisteredCountry":{"GeoNameID":0,"IsoCode":"","Names":null},"RepresentedCountry":{"GeoNameID":0,"IsoCode":"","Names":null,"Type":""},"Subdivisions":null,"Traits":{"IsAnonymousProxy":false,"IsSatelliteProvider":false}}%
```

This is probably because the geolookup service is seeing the proxy's IP and not the peer's, but that's also odd, it should have given me the NL coordinates then. 

I verified that the geolookup service supports `X-Forwarded-For` so maybe be the proxy is not crafting a correct `X-Forwarded-For`? Do you know how can we test that one, @aranhoide?

**Comment by myleshorton** [comment]
Yeah our proxies are definitely not forwarding ```X-Forwarded-For```, @xiam. Sorry I meant to mention that. They don't add it in order to add some degree of added privacy for Lantern users so sites don't know their IP address, but it also means that for now we have to use direct domain fronting for the geo lookup and also for the config lookup for the same reason (since the config lookup bases the config on your geolocation).
